### ISSUE TITLE
FS_FILE_TYPE information in snapshot file diff output

### ISSUE THREAD

**Issue by pallabpain** [Qumulo/qumulo-api-introduction#2]
I've been playing around with Qumulo APIs for a PoC that I'm working on and while trying out the snapshot file diff API, I realized that the information is pretty limited.
```
{'entries': [
             {'op': 'CREATE', 'path': '/share01/OZCR8KQ/BVPGWYDXW/O13ZHC'},

             {'op': 'CREATE',

              'path': '/share01/OZCR8KQ/BVPGWYDXW/25113CHQ/0FQ4K0EFF/DEITSA3D2W'},

             {'op': 'CREATE',

              'path': '/share01/OZCR8KQ/BVPGWYDXW/25113CHQ/0FQ4K0EFF/KUBPSK'},

             {'op': 'CREATE',

              'path': '/share01/OZCR8KQ/BVPGWYDXW/25113CHQ/0FQ4K0EFF/JV0TR'},

             {'op': 'CREATE',

              'path': '/share01/OZCR8KQ/BVPGWYDXW/25113CHQ/0FQ4K0EFF/6LLYLH'}],

 'paging': {'next': '/v2/snapshots/20/changes-since/19?after=AAAAAAAAAAACAAAAoYcKAAAAAAAAAAAAp38LAAYAAAA2TExZTEgTAAAAFAAAAA==&limit=5'}}
```
I only get the operation and the path to the file.

But I also want to know the type of file. Whether it's a file or a directory or a symlink or socket, etc. I also found out that there are APIs that give information about the files in a fs_path. But from a pure API consumer point-of-view, have to gather information from two different APIs just isn't quite feasible in terms of performance when there are tens of millions of files and the diff itself is in millions.

I did manage to get the information from a tree walk on the path and creating a map from it. Then iterating over the diff and then finding out the type. But building the entire file-type map isn't a feasible operation.

Can the **type** information be made a part of this output? For example,
```
 {'op': 'CREATE',
 'path': '/share01/OZCR8KQ/BVPGWYDXW/25113CHQ/0FQ4K0EFF/KUBPSK'
 'type': 'FS_FILE_TYPE_DIRECTORY'},

{'op': 'CREATE',
 'path': '/share01/OZCR8KQ/BVPGWYDXW/25113CHQ/0FQ4K0EFF/JV0TR'
 'type': 'FS_FILE_TYPE_FILE'},

{'op': 'CREATE',
 'path': '/share01/OZCR8KQ/BVPGWYDXW/25113CHQ/0FQ4K0EFF/6LLYLH'
 'type': 'FS_FILE_TYPE_FILE'}
```

**Comment by pallabpain** [comment]
On second thought, I've observed that all directories have a trailing "/" in the path. And hence,
```
if path[-1] == "/":
    path is a directory
else:
    path is not a directory 
```
Is this a correct assumption?

**Comment by TommyUnger** [comment]
Thanks for the idea, Pallab. Please reach out to your Sales Rep at Qumulo for this feature request.

For what it's worth, Qumulo's API could handle millions of additional requests. It scales well and I've been able to get as many as 10,000 metadata reads per second from a single cluster with parallelized workloads.

**Comment by TommyUnger** [comment]
Yes, you can also use your trailing slash approach. It won't help you for all of the non-directory types though, so watch out for those.

**Comment by pallabpain** [comment]
Thanks, Tommy. I'll try and reach out via proper channels for the feature request once I have a working PoC to demonstrate.

I'm making it work with the trailing slash approach for now since I only require the distinction between a directory and a non-directory. I did try gathering information on a file/dir from the various APIs that are available. However, that approach makes the overall diff generation slow since it took time to generate the entire tree using `fs.tree_walk_preorder` or `fs.tree_walk_postorder` before I started fetching the diff and extracting the metadata from the tree. Having all the required information in the `snapshot.get_all_snapshot_tree_diff` response certainly helps to speed up things a lot while performing incremental backups.

Also, the `snapshot.get_all_snapshot_tree_diff` method returns the overall diff list in post-order. It will be great if we can choose to have the method send us a pre-order form of the list.

Otherwise, the overall API performance has been great!
